pytorch_lightning/profiler/pytorch.py

```diff
@@ -448,68 +613,242 @@ def on_trace_ready(profiler):
+
+    def _delete_profiler(self) -> None:
         if self.profiler is not None:
-            self.profiler.__exit__(None, None, None)
+            if not _KINETO_AVAILABLE or self._emit_nvtx:
+                self.profiler.__exit__(None, None, None)
+            else:
+                # At the moment, the profiler should already be stoped.
+                if torch.autograd._profiler_enabled():
+                    self._transit_profiler(KinetoProfilerState.NONE)
+                    self._save_result(self._start_action_name)
             self._cache_functions_events()
             self.profiler = None
 
-        if self._schedule is not None:
-            self._schedule.reset()
+    def _save_result(self, action_name: str) -> None:
+        os.makedirs(self.dirpath, exist_ok=True)
+        if self._export_to_chrome:
+            path = os.path.join(
+                self.dirpath,
+                "{}.{}.pt.trace.json".format(
+                    self._prepare_filename(action_name=action_name, extension=""), int(time.time() * 1000)
+                ),
+            )
+            self.profiler.export_chrome_trace(path)
+        if self._export_to_flame_graph:
+            path = os.path.join(self.dirpath, self._prepare_filename(action_name=action_name, extension=".stack"))
+            self.profiler.export_stacks(path, metric=self._metric)
 
-        if self._parent_profiler is not None:
-            self._parent_profiler.__exit__(None, None, None)
-            self._parent_profiler = None
+    @property
+    def profile_steps(self):
+        return self._wait_step + self._warmup_step + self._active_step
 
-        if self._register is not None:
-            self._register.__exit__(None, None, None)
-            self._register = None
+    def get_total_steps(self, action_name) -> int:
+        trainer = self._lightning_module.trainer
 
-    def teardown(self, stage: Optional[str] = None) -> None:
-        self._delete_profilers()
+        if action_name.endswith("training_step"):
+            return trainer.num_training_batches
+        if action_name.endswith("validation_step"):
+            return sum(trainer.num_val_batches) + sum(trainer.num_sanity_val_batches)
+        if action_name.endswith("test_step"):
+            return sum(trainer.num_test_batches)
+        if action_name.endswith("predict_step"):
+            return sum(trainer.num_predict_batches)
 
-        for k in list(self._recording_map):
-            self.stop(k)
-        self._recording_map = {}
+    def _transit_profiler(self, new_state: KinetoProfilerState):
+        action_list = self._action_map.get((self._profiler_state, new_state))
+        if action_list:
+            for action in action_list:
+                action(self.profiler)
+            self._profiler_state = new_state
 
-        super().teardown(stage=stage)
+
+if _TORCH_GREATER_EQUAL_1_9 and hasattr(torch.profiler, "_KinetoProfile"):
```

Comment:
Isn't Kineto part of 1.8.1?

Reply:
yes, just not want to import _TORCH_GREATER_EQUAL_1_8_1 any more.